### DISCUSSION TITLE
Jupyter kernel dies every time when I plot something!

### DISCUSSION THREAD

**Discussion by minsikseo-cdl** [pyvista/pyvista#1847] [Q&A]
Hi, I just tried [Jupyter Notebook Plotting](https://docs.pyvista.org/user-guide/jupyter/index.html) as below:

```
import pyvista as pv
from pyvista import examples

mesh = examples.download_lucy()
mesh.plot(color='lightgrey', pbr=True, metallic=0.2,
          jupyter_backend='pythreejs')
```

Then the Jupyter kernel just dies and restarts every time without any error message.

I've installed ```pyvista``` by using ```conda install -c conda-forge pyvista```, so I thought there would not be any dependency problem.
I tried it on Jupiter Lab, not on Jupiter notebook. Can this be a problem?

The attached file is my environment.
[myenv.txt](https://github.com/pyvista/pyvista/files/7592412/myenv.txt)

**Comment by banesullivan** [comment]
Thanks for attaching your environment file, that is helpful. Would you please also add a report (disabling the GPU reporting since that plots something in the background):

```py
>>> import pyvista
>>> pyvista.Report(gpu=False)
```

(And also with `gpu=True` if that works for you).

-----

Do you have a GPU on your machine? If not, you may have the wrong version of VTK as from your env file, I see you have:

```
vtk                       9.0.1           no_osmesa_py37h111716f_107    conda-forge
```

You may need the `with_osmesa` from conda forge.

**Reply by minsikseo-cdl** [reply]
Thank you for the answer!
Yes I have 4 GPUs on my machine. But, somehow, I think it's probably the GPUs are in trouble. 

When I enter ```nvidia-smi``` in terminal, it gives
```
Fri Nov 26 06:25:47 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.82.00    Driver Version: 470.82.00    CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:3F:00.0 Off |                  N/A |
| 39%   37C    P0   103W / 350W |      0MiB / 24268MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce ...  Off  | 00000000:40:00.0 Off |                  N/A |
| 40%   37C    P0   114W / 350W |      0MiB / 24268MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA GeForce ...  Off  | 00000000:62:00.0 Off |                  N/A |
| 42%   36C    P0   104W / 350W |      0MiB / 24268MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA GeForce ...  Off  | 00000000:63:00.0 Off |                  N/A |
| 43%   36C    P0   107W / 350W |      0MiB / 24268MiB |      3%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

I'm working in a docker container with the same image that is well-functioning when I use ```pytorch```. But, ```torch.cuda.is_available()``` gives ```False```. Likewise, ```pyvista.Report(gpu=True)``` also kills the kernel while ```pyvista.Report(gpu=False)``` gives:
![image](https://user-images.githubusercontent.com/80688256/143537707-430572ee-f2ac-4f6f-9d4d-a2b05bb0032b.png)

I installed a bunch of packages after ```run``` the container, so it might cause something wrong. I'll rerun a new container and try with ```pyvista``` alone, then I'll let you know the results.

**Reply by minsikseo-cdl** [reply]
@banesullivan 
uh-oh... GPUs get unavailable right after I install ```pyvista```.

Actually, I have one more thing to ask: Doesn't it automatically resolve the dependency problem to install using ```conda```? I had to manually install ```libgl1-mesa-glx``` and ```mesalib```.

**Reply by banesullivan** [reply]
Oh wow, this might be a wormhole... and I don't know how much I will be able to help

> I'm working in a docker container

AFAIK, you'll need osmesa support then because you'll be rendering in a *headless environment*. Maybe tomorrow I can hop on my linux machine with an NVIDIA GPU and try to build a docker image that can access the GPU with PyVista.

>  Doesn't it automatically resolve the dependency problem to install using conda? I had to manually install libgl1-mesa-glx and mesalib.

It should... but VTK is always a wild card, even with conda

**Reply by banesullivan** [reply]
Can you share any parts of your Dockerfile?

**Reply by minsikseo-cdl** [reply]
@banesullivan
Of course, this is my docker file

```
FROM minsik/torch:cu111
# Install prerequisites
RUN apt update && apt upgrade -y && \
        apt install -y vim git wget curl fish libsuitesparse-dev \
        libblas-dev liblapack-dev libopenmpi-dev openmpi-bin && chsh -s /usr/bin/fish
RUN conda install --name base -y -c conda-forge \
        scikit-umfpack cyipopt scipy pytorch-lightning matplotlib numba jupyterlab joblib cython mpi4py
RUN export PYTHONPATH=${PYTHONPATH}:${HOME}/git/paropt && \
        export PAROPT_DIR=${HOME}/git/paropt && \
        PATH=$PATH:$PAROPT_DIR
RUN cd ~ && mkdir git && cd git && git clone https://github.com/smdogroup/paropt.git && \
        cd paropt && cp Makefile.in.info Makefile.in && make && \
        python setup.py install && make interface
# Install code-server (Optional)
RUN curl -fsSL https://code-server.dev/install.sh | sh && code-server --install-extension ms-python.python
CMD code-server --bind-addr 0.0.0.0:8080
```

**Comment by dkobayas-cyber** [comment]
I'm facing the same problem with my JupyterLab notebook. The only difference is that I'm not using Docker. I tried the following: 
```python 

import pyvista as pv
pv.set_jupyter_backend('trame')

sphere = pv.Sphere()

plotter = pv.Plotter(notebook=True)
plotter.add_mesh(sphere)
plotter.show()

```
, which killed my kernel. I used to have a similar problem when I used `opencv.imshow()` on the JupyterLab, which may or may not be relevant to this problem. 

`pyvista.Report(gpu=True)` gives the following: 
![image](https://user-images.githubusercontent.com/85417498/233506764-7113bc3e-7e92-4869-81bd-746f0123f6a4.png) 

I also attached the environment file below. I installed PyVista by `conda install -c conda-forge pyvista`. 
 
[environment.txt](https://github.com/pyvista/pyvista/files/11291081/environment.txt)

**Reply by dkobayas-cyber** [reply]
I tried running the same code on JupyterNotebook, which also crushed the kernel but gave me more detailed error messages: 
```
/depot/cfrueh/apps/env_pyvista/lib/python3.8/site-packages/pyvista/plotting/plotting.py:147: UserWarning: 
This system does not appear to be running an xserver.
PyVista will likely segfault when rendering.

Try starting a virtual frame buffer with xvfb, or using
  ``pyvista.start_xvfb()``

  warnings.warn(
/depot/cfrueh/apps/env_pyvista/lib/python3.8/site-packages/pyvista/jupyter/notebook.py:58: UserWarning: Failed to use notebook backend: 

Your build of VTK does not have the proper web modules enabled.
These modules are typically enabled by default with the
`-DVTK_GROUP_ENABLE_Web:STRING=WANT` build flag.

Conda users: This is a known issue with the conda-forge VTK feedstock.
See https://github.com/conda-forge/vtk-feedstock/pull/258


Falling back to a static output.
  warnings.warn(
```
It seems like I need to build VTK with a proper web module (?), which I'm not familiar with. I also visited the link in the error message but I don't have a clear idea of how to fix this problem. Please give me some help!

**Reply by banesullivan** [reply]
@dkobayas-cyber, please upgrade to a more recent version of VTK (that is why you see the messages about the missing web module). You have VTK 9.0.3 (~3 years old); please upgrade to 9.2.x to leverage the latest Jupyter Notebook plotting capabilities.

As for your kernel crashing, this sounds like you do not have libgl configured/installed for your system (independent of VTK/PyVista) -- hence why you also have issues with `opencv`.

Can you open a terminal to attempt to plot with PyVista and see if any error messages are given on segfault? This will help track down what exactly needs to be done.

**Reply by dkobayas-cyber** [reply]
@banesullivan, Thank you for your heads-up. I have re-created an environment with VTK 9.2.6. Attached is the list of packages in the new environment: [environment_new.txt](https://github.com/pyvista/pyvista/files/11297542/environment_new.txt). I realized that the conda installation by `conda install -c conda-forge pyvista` makes me install the VTK 9.0.3, which was why I had this old version. 

Below are the results of the code and errors on the terminal: 
```ipython
dkobayas@gilbreth-e014:~ $ ipython
Python 3.8.5 (default, Sep  4 2020, 07:30:14) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import pyvista as pv

In [2]: mesh = pv.Sphere()

In [3]: mesh.points
Out[3]: 
pyvista_ndarray([[-5.5511151e-17,  0.0000000e+00, -5.0000000e-01],
                 [ 5.5511151e-17,  0.0000000e+00,  5.0000000e-01],
                 [-5.4059509e-02,  0.0000000e+00, -4.9706897e-01],
                 ...,
                 [-1.5616201e-01, -3.3193260e-02,  4.7382659e-01],
                 [-1.0513641e-01, -2.2347433e-02,  4.8831028e-01],
                 [-5.2878179e-02, -1.1239604e-02,  4.9706897e-01]],
                dtype=float32)

In [4]: mesh.plot(background='w', show_edges=True)
/depot/cfrueh/apps/env_pyvista/lib/python3.8/site-packages/pyvista/plotting/plotting.py:147: UserWarning: 
This system does not appear to be running an xserver.
PyVista will likely segfault when rendering.

Try starting a virtual frame buffer with xvfb, or using
  ``pyvista.start_xvfb()``

  warnings.warn(
Segmentation fault (core dumped)
```
As you mentioned, I am getting the segmentation fault error.

**Reply by banesullivan** [reply]
I think you're missing system GL libraries... maybe `libgl1-mesa-dev` is what you need? `sudo apt install libgl1-mesa-dev`

**Reply by dkobayas-cyber** [reply]
@banesullivan After some tests, we found that the latest vtk 9.2.6 version has a system glibc dependency issue. After I downgrade it to vtk version 9.2.2, the problem of kernel crushing has been resolved. However, the plot is still not displayed correctly. 
* The code I tested
```python 
%matplotlib inline
from pyvista import set_plot_theme
set_plot_theme("document")

import numpy as np
import pyvista as pv
from pyvista import examples

mesh = examples.download_bunny_coarse()

pl = pv.Plotter()
pl.add_mesh(mesh, show_edges=True, color="white")
pl.add_points(mesh.points, color="red", point_size=10)
pl.camera_position = [(0.02, 0.30, 0.73), (0.02, 0.03, -0.022), (-0.03, 0.94, -0.34)]
pl.show()
```
* Without ipywidgets, I could display the bunny but the plot was static. I got a warning messaging saying that the ipywidget is not installed. 
* After installing ipywidgets, I became unable to display the bunny and got an error message: "localhost refused to connect."

**Reply by ggrrll** [reply]
downgrading VTK did not solve te issue for me :/

**Reply by banesullivan** [reply]
@ggrrll, please open a new discussion or issue with more context into the problem you're experiencing.

@dkobayas-cyber:

> After installing ipywidgets, I became unable to display the bunny and got an error message: "localhost refused to connect."

It sounds like you're using Jupyter on some remote service that isn't on `localhost`. This has been tricky for some users to set up but there are a handful of issues, and documentation outlining the process. For starter's reference:

- https://docs.pyvista.org/version/stable/user-guide/jupyter/trame.html#jupyter-server-proxy
- https://github.com/pyvista/pyvista/issues/3979

**Reply by ggrrll** [reply]
sorry -- I actually solved the dead kernel issue by installing `libgl1-mesa-glx`

**Comment by chams333** [comment]
I added the following code at the beginning of the notebook, and the crashes stopped. 

`import pyvista

from pyvista import demos
pyvista.start_xvfb()
demos.plot_wave()
`
Somehow, there is an issue in the order at which pyvista and dolfinx are called. If I try to run pyvista after I load dolfinx, crashes occur.

**Reply by banesullivan** [reply]
Reminds me of https://github.com/pyvista/pyvista/issues/4462

It sounds like dolfinx does something incompatible to the environment

**Reply by Irratzo** [reply]
Thank you, adding `pyvista.start_xvfb()` resolved the issue for me. Sometimes I have to restart the kernel, but then it works, with @chams333  minimal working example (MWE) above, and with my own plotting code. My system is Apple Silicon, MacOS 15.1.1, Python env venv (uv) and conda (micromamba) (both behave identically, same package versions).

```bash
$ python --version
Python 3.12.3

$ uv pip list | grep -e "pyvista" -e "vtk" -e "trame"
pyvista                   0.44.1
trame                     3.7.0
trame-client              3.5.0
trame-server              3.2.3
trame-vtk                 2.8.12
trame-vuetify             2.7.2
vtk                       9.3.1

$ uv pip list | grep -e "ipykernel" -e "jupyterlab"
ipykernel                 6.29.5
jupyterlab                3.6.8
jupyterlab-pygments       0.3.0
jupyterlab-server         2.27.3
jupyterlab-widgets        3.0.13
```
